src/librustc/infer/resolve.rs

```diff
@@ -29,7 +29,7 @@ impl<'a, 'tcx> TypeFolder<'tcx> for OpportunisticVarResolver<'a, 'tcx> {
     }
 
     fn fold_ty(&mut self, t: Ty<'tcx>) -> Ty<'tcx> {
-        if !t.has_infer_types() {
+        if !t.has_infer_types() && !t.has_infer_consts() {
```

Comment:
It might be a good idea if we audit other uses of `has_infer_types` to make sure they don't also need to check `has_infer_consts`…